src/components/shared/SpentTime/SpentTime.tsx

```diff
@@ -0,0 +1,128 @@
+
+import styles from "./SpentTime.module.scss";
+
+interface SpentTimeProps {
+  userId: UserId;
+}
+
+export const SpentTime: React.FC<SpentTimeProps> = ({ userId }) => {
+  const { isLoaded: isWorldLoaded, spaceId } = useWorldAndSpaceByParams();
+  const [visits, setVisits] = useState<WithId<UserVisit>[]>([]);
+  const [venues, setVenues] = useState<WithId<AnyVenue>[]>([]);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+
+  const firestore = useFirestore();
+
+  const fetchAllVenues = useCallback(async () => {
+    const userSnapshot = await getUserRef(userId).get();
+    const visitsSnapshot = await userSnapshot.ref.collection("visits").get();
+
+    const visits: WithId<UserVisit>[] =
+      visitsSnapshot.docs.map(
+        (visitSnapshot) =>
+          ({
+            ...visitSnapshot.data(),
+            id: visitSnapshot.id,
+          } as WithId<UserVisit>)
+      ) ?? [];
+
+    const venuesRequests = chunk(
+      visits,
+      FIRESTORE_QUERY_IN_ARRAY_MAX_ITEMS
+    ).map((visitChunk) =>
+      getDocs(
+        query(
+          collection(firestore, COLLECTION_SPACES),
+          where(
+            "name",
+            "in",
+            visitChunk.map((visit) => visit.id)
+          )
+        )
+      )
+    );
+
+    let venues: WithId<AnyVenue>[] = [];
+    const hasVenuesRequests = isTruthy(venuesRequests);
+
+    // If there are no venues visited avoid sending the request.
+    if (hasVenuesRequests) {
+      const requestSnapshots = await Promise.all(venuesRequests);
+
+      // Promise all returns arrays as response. That's why there is so much depth.
+      // TODO: Same logic can be used for the private chats as well.
+      venues = requestSnapshots.flatMap((venuesSnapshot) =>
+        venuesSnapshot.docs.map(
+          (venueSnapshot) =>
+            ({
+              ...venueSnapshot.data(),
+              id: venueSnapshot.id,
+            } as WithId<AnyVenue>)
+        )
+      );
+    }
+
+    setVenues(venues);
+    setVisits(visits);
+  }, [firestore, userId, setVenues, setVisits]);
+
+  useEffect(() => {
+    setIsLoading(true);
+    const stop = () => setIsLoading(false);
+
+    fetchAllVenues()
+      .catch((e) => console.error(SpentTime.name, e))
+      .finally(stop);
+
+    return stop;
+  }, [setIsLoading, fetchAllVenues]);
+
+  const venueNames = useMemo(() => venues.map((venue) => venue.name), [venues]);
+
+  const visitHours = useMemo(() => {
+    if (!visits) return 0;
+
+    const visitSeconds = visits
+      .filter((v) => v.id === spaceId || venueNames.includes(v.id))
+      .reduce((acc, v) => acc + v.timeSpent, 0);
+    return Math.floor(visitSeconds / (60 * 60));
+  }, [visits, spaceId, venueNames]);
+
+  if (isLoading) {
+    return (
+      <div data-bem="SpentTime--loading" className={styles.general}>
+        Loading...
+      </div>
+    );
+  }
+
+  if (!isWorldLoaded) {
+    return null;
+  }
```

Comment:
These can be done similar to https://github.com/sparkletown/sparkle/blob/redesign-2021/src/components/templates/ConversationSpace/ConversationSpace.hoc.tsx and maybe that would be a good start for the elsewhere comment above

Reply:
`isLoading` is actively used in the component, if we move all the logic to hook, it'll be almost the same as the component now but without the HTML. Once we rewrite the logic of calculating online hours, then we can refactor that I think.